radixdlt-core/radixdlt/src/main/java/com/radixdlt/network/p2p/addressbook/PeerPostponingManager.java

```diff
@@ -0,0 +1,170 @@
+ * shall be liable for any direct or indirect, special, incidental, consequential or other
+ * losses of any kind, in tort, contract or otherwise (including but not limited to loss
+ * of revenue, income or profits, or loss of use or data, or loss of reputation, or loss
+ * of any economic or other opportunity of whatsoever nature or howsoever arising), arising
+ * out of or in connection with (without limitation of any use, misuse, of any ledger system
+ * or use made or its functionality or any performance or operation of any code or protocol
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+package com.radixdlt.network.p2p.addressbook;
+
+import com.radixdlt.network.p2p.NodeId;
+import com.radixdlt.network.p2p.RadixNodeUri;
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Optional;
+
+/**
+ * Maintain information about peers connections to which should be postponed.
+ *
+ * <p>Class implements simple exponential backoff strategy, doubling postpone time for every
+ * recorded failure.
+ */
+final class PeerPostponingManager {
+  private final Map<RadixNodeUri, PostponeInfo> postponedPeers = new HashMap<>();
+  private final Clock clock;
+  private final Duration initialPostpone;
+  private final Duration maxPostpone;
+
+  public PeerPostponingManager(Clock clock, Duration initialPostpone, Duration maxPostpone) {
+    this.clock = clock;
+    this.initialPostpone = initialPostpone;
+    this.maxPostpone = maxPostpone;
+  }
+
+  public boolean recordFailure(RadixNodeUri uri) {
```

Comment:
Ditto here - we have a method here with a side effect and a returned boolean. How would you feel about separating the side effect (record failure) from the question (shouldRemoveFromAddressBookDueToFailures?)

Reply:
Here I had a choice between action/query separation and creation of potential space for race condition. Given how usually hard to nail down and fix race conditions, I prefer to live with atomic action-query. But we may discuss naming :)

Reply:
I don't necessarily see a problem with lack of atomicity here, as long as `shouldRemoveFromAddressBookDueToFailures` doesn't make any assumption that the node exists in `postponedPeers`. Indeed, if a second successful request comes in whilst processing a first unsuccessful request, it'd be better if the first request _didn't_ remove the address from the address book! aha. But it's probably quite unlikely.

Alternatively, I'm happy to keep the current structure - but in terms of naming, my personal preferences (from imo most honest to least honest) are 1. Naming both the side effect and return `recordFailureAndReturnIfShouldBeKeptInAddressBook`, or 2. Naming just the side effect (if you have to) - in which case `recordFailure` is okay, before 3. Naming just the return (as it's easy to miss the side effect).